cherry-pick 1.1: log: ensure panic details are logged to file even when stderr is not redirected

knz opened the pull request:
Cherry-pick of #20839.

Prior to this patch, panic details would not be captured to logs
unless the log file was forcing redirection of stderr. This was
notoriously not the case when passing `--logtostderr` or
`--logtostderr=INFO`.  (In contrast, `--logtostderr=WARNING` would
redirect stderr and cause the panic details to be properly captured.)

This patch fixes the issue by capturing the panic details every time
stderr is not redirected.

Release note (bug fix): the crash details are now properly copied to
the log file when starting a node with `--logtostderr` (and in some
other circumstances where they could be lost previously).

cc @cockroachdb/release

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20842)
<!-- Reviewable:end -->

knz committed:
log: ensure panic details are logged to file even when stderr is not redirected

Prior to this patch, panic details would not be captured to logs
unless the log file was forcing redirection of stderr. This was
notoriously not the case when passing `--logtostderr` or
`--logtostderr=INFO`.  (In contrast, `--logtostderr=WARNING` would
redirect stderr and cause the panic details to be properly captured.)

This patch fixes the issue by capturing the panic details every time
stderr is not redirected.

Release note (bug fix): the crash details are now properly copied to
the log file when starting a node with `--logtostderr` (and in some
other circumstances where they could be lost previously).

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20842#-:-L0flLUYVEwopnK250eT:bnfp4nl)*
<!-- Sent from Reviewable.io -->
